### PR TITLE
[Package Signing] Add Trusted Timestamps to Validation Context

### PR DESCRIPTION
**Pull request by loic-sharma**
Each signature on a package is stamped with a signed timestamp from a trusted source. This will be represented by the `TrustedTimestamp` entity. Each `PackageSignature` has one `TrustedTimestamp`. Each `TrustedTimestamp` was signed with one `Certificate`. A `Certificate` may be depended on 0 or more `PackageSignature`s as well as 0 or more `TrustedTimestamp`s.

Miscelleanous changes piggybacking this pull request:
* Added an `Unknown` value to the `PackageSignatureStatus` enum. This will be necessary to distinguish how an invalidated package should be handled (specifically, should we alert on this failed signature validation?). I updated the spec to reflect this new enum value.
* Enforced a default value for the `PackageSignature`'s `CreatedAt` property.

Once this pull request is merged, dev will be merged into master and new packages will be created for the package signing branch!

## SQL Schema

```
CREATE TABLE [dbo].[ValidatorStatuses] (
    [ValidationId] [uniqueidentifier] NOT NULL,
    [PackageKey] [int] NOT NULL,
    [ValidatorName] [nvarchar](max) NOT NULL,
    [State] [int] NOT NULL,
    [RowVersion] rowversion NOT NULL,
    CONSTRAINT [PK_dbo.ValidatorStatuses] PRIMARY KEY ([ValidationId])
)
CREATE INDEX [IX_ValidatorStatuses_PackageKey] ON [dbo].[ValidatorStatuses]([PackageKey])

IF schema_id('signature') IS NULL
    EXECUTE('CREATE SCHEMA [signature]')
CREATE TABLE [signature].[Certificates] (
    [Key] [bigint] NOT NULL IDENTITY,
    [Thumbprint] [char](20) NOT NULL,
    [Status] [int] NOT NULL,
    [StatusUpdateTime] [datetime2](7),
    [NextStatusUpdateTime] [datetime2](7),
    [LastVerificationTime] [datetime2](7),
    [RevocationTime] [datetime2](7),
    [ValidationFailures] [int] NOT NULL,
    [RowVersion] rowversion NOT NULL,
    CONSTRAINT [PK_signature.Certificates] PRIMARY KEY ([Key])
)
CREATE UNIQUE INDEX [IX_Certificates_Thumbprint] ON [signature].[Certificates]([Thumbprint])
CREATE TABLE [signature].[PackageSignatures] (
    [Key] [bigint] NOT NULL IDENTITY,
    [PackageKey] [int] NOT NULL,
    [CertificateKey] [bigint] NOT NULL,
    [CreatedAt] [datetime2](7) NOT NULL,
    [Status] [int] NOT NULL,
    CONSTRAINT [PK_signature.PackageSignatures] PRIMARY KEY ([Key])
)
CREATE INDEX [IX_PackageSignatures_PackageKey] ON [signature].[PackageSignatures]([PackageKey])
CREATE INDEX [IX_PackageSignatures_CertificateKey] ON [signature].[PackageSignatures]([CertificateKey])
CREATE INDEX [IX_PackageSignatures_Status] ON [signature].[PackageSignatures]([Status])
CREATE TABLE [signature].[PackageSigningStates] (
    [PackageKey] [int] NOT NULL,
    [PackageId] [nvarchar](128) NOT NULL,
    [PackageNormalizedVersion] [nvarchar](64) NOT NULL,
    [SigningStatus] [int] NOT NULL,
    CONSTRAINT [PK_signature.PackageSigningStates] PRIMARY KEY ([PackageKey])
)
CREATE INDEX [IX_PackageSigningStates_PackageId_PackageNormalizedVersion] ON [signature].[PackageSigningStates]([PackageId], [PackageNormalizedVersion])
CREATE TABLE [signature].[TrustedTimestamps] (
    [Key] [bigint] NOT NULL IDENTITY,
    [PackageSignatureKey] [bigint] NOT NULL,
    [CertificateKey] [bigint] NOT NULL,
    [Value] [datetime2](7) NOT NULL,
    CONSTRAINT [PK_signature.TrustedTimestamps] PRIMARY KEY ([Key])
)
CREATE INDEX [IX_PackageSignatureKey] ON [signature].[TrustedTimestamps]([PackageSignatureKey])
CREATE INDEX [IX_CertificateKey] ON [signature].[TrustedTimestamps]([CertificateKey])
CREATE TABLE [signature].[CertificateValidations] (
    [Key] [bigint] NOT NULL IDENTITY,
    [CertificateKey] [bigint] NOT NULL,
    [ValidationId] [uniqueidentifier] NOT NULL,
    [Status] [int],
    CONSTRAINT [PK_signature.CertificateValidations] PRIMARY KEY ([Key])
)
CREATE INDEX [IX_CertificateValidations_CertificateKey_ValidationId] ON [signature].[CertificateValidations]([CertificateKey], [ValidationId])
CREATE INDEX [IX_CertificateValidations_ValidationId] ON [signature].[CertificateValidations]([ValidationId])
ALTER TABLE [signature].[PackageSignatures] ADD CONSTRAINT [FK_signature.PackageSignatures_signature.Certificates_CertificateKey] FOREIGN KEY ([CertificateKey]) REFERENCES [signature].[Certificates] ([Key]) ON DELETE CASCADE
ALTER TABLE [signature].[PackageSignatures] ADD CONSTRAINT [FK_signature.PackageSignatures_signature.PackageSigningStates_PackageKey] FOREIGN KEY ([PackageKey]) REFERENCES [signature].[PackageSigningStates] ([PackageKey]) ON DELETE CASCADE
ALTER TABLE [signature].[TrustedTimestamps] ADD CONSTRAINT [FK_signature.TrustedTimestamps_signature.Certificates_CertificateKey] FOREIGN KEY ([CertificateKey]) REFERENCES [signature].[Certificates] ([Key]) ON DELETE CASCADE
ALTER TABLE [signature].[TrustedTimestamps] ADD CONSTRAINT [FK_signature.TrustedTimestamps_signature.PackageSignatures_PackageSignatureKey] FOREIGN KEY ([PackageSignatureKey]) REFERENCES [signature].[PackageSignatures] ([Key]) ON DELETE CASCADE
ALTER TABLE [signature].[CertificateValidations] ADD CONSTRAINT [FK_signature.CertificateValidations_signature.Certificates_CertificateKey] FOREIGN KEY ([CertificateKey]) REFERENCES [signature].[Certificates] ([Key]) ON DELETE CASCADE
```

### CODE REVIEW

File: src/NuGet.Services.Validation/Entities/PackageSignature.cs [hunk 1]
```diff
@@ -21,22 +21,14 @@ public class PackageSignature
...
-        /// The key to the end <see cref="Certificate"/> used to create this package signature.
+        /// The key to the end <see cref="Validation.Certificate"/> used to create this package signature.
         /// </summary>
         public long CertificateKey { get; set; }
 
         /// <summary>
-        /// The time at which this signature was created. A signature is valid as long as it was signed
-        /// before its certificates were revoked and/or expired. This timestamp MUST come from a trusted
-        /// timestamp authority.
+        /// The time at which this record was created, used to detect signatures that are stuck "InGracePeriod".
```
Comment (scottbommarito) [inline src/NuGet.Services.Validation/Entities/PackageSignature.cs]: nit: `<see cref="PackageSignatureStatus.InGracePeriod"/>`

File: src/NuGet.Services.Validation/Entities/TrustedTimestamp.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace NuGet.Services.Validation
+{
+    /// <summary>
+    /// A timestamp created by a <see href="https://tools.ietf.org/html/rfc3161">Trusted Timestamp Authority</see>.
```
Comment (scottbommarito) [inline src/NuGet.Services.Validation/Entities/TrustedTimestamp.cs]: had no idea you could do `href`s in here, awesome

File: src/NuGet.Services.Validation/Entities/TrustedTimestamp.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    /// </summary>
+    public class TrustedTimestamp
+    {
+        /// <summary>
+        /// The key to the <see cref="PackageSignature"/> that depends on this trusted timestamp.
+        /// </summary>
+        public long PackageSignatureKey { get; set; }
```
Comment (joelverhagen) [inline src/NuGet.Services.Validation/Entities/TrustedTimestamp.cs]: Will it ever be possible to have multiple timestamps per package signature?

**Comment by joelverhagen** [comment]
Could you provide the new SQL?

File: src/NuGet.Services.Validation/Entities/TrustedTimestamp.cs (hunk 3 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation/Entities/TrustedTimestamp.cs]: Excellent question. Currently, the spec mandates a timestamp, but it doesn't seem to require EXACTLY one timestamp.

@dtivel Could we strengthen the wording in the spec about this?

Comment (joelverhagen) [reply src/NuGet.Services.Validation/Entities/TrustedTimestamp.cs]: We can move forward with this either way by giving `TrustedTimestamp` its own key.

**Comment by loic-sharma** [comment]
Addressed feedback; provided generated SQL migration.

File: src/NuGet.Services.Validation/Entities/TrustedTimestamp.cs (hunk 3 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation/Entities/TrustedTimestamp.cs]: I went ahead and made this be a one-to-many relationship as well. In the initial version of package signing, each package should have ONE signature with ONE trusted timestamp authority. However, this schema is flexible enough to give room for change in the future :)

File: src/NuGet.Services.Validation/Entities/PackageSignature.cs (hunk 1 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation/Entities/PackageSignature.cs]: Fixed!

**Pull request merged**